examples/microViT.py

```diff
@@ -0,0 +1,237 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+# CREDITS:
+# inspired by
+# https://github.com/nateraw/lightning-vision-transformer
+# which in turn references https://github.com/lucidrains/vit-pytorch
+
+# Orignal author: Sean Naren
+
+import math
+
+import pytorch_lightning as pl
+import torch
+from pl_bolts.datamodules import CIFAR10DataModule
+from torch import nn
+from torchmetrics import Accuracy
+from torchvision import transforms
+
+from xformers.factory import xFormer, xFormerConfig
+
+
+class VisionTransformer(pl.LightningModule):
+    def __init__(
+        self,
+        steps,
+        learning_rate=1e-4,
+        weight_decay=0.0001,
+        image_size=32,
+        num_classes=10,
+        patch_size=4,
+        dim=256,
+        n_layer=12,
+        n_head=4,
+        resid_pdrop=0.1,
+        attn_pdrop=0.1,
+        mlp_pdrop=0.1,
+        attention="scaled_dot_product",
+        hidden_layer_multiplier=4,
+        linear_warmup_ratio=0.05,
+        classifier="gap",
+    ):
+
+        super().__init__()
+
+        self.save_hyperparameters()
+
+        self.learning_rate = learning_rate
+        self.weight_decay = weight_decay
+        self.steps = steps
+        self.linear_warmup_ratio = linear_warmup_ratio
+
+        assert image_size % patch_size == 0
+
+        num_patches = (image_size // patch_size) ** 2
+        self.dim = dim
+        self.classifier = classifier
+
+        # A list of the encoder or decoder blocks which constitute the Transformer.
+        xformer_config = [
+            {
+                "block_config": {
+                    "block_type": "encoder",
+                    "num_layers": self.hparams.n_layer,
+                    "dim_model": self.hparams.dim,
+                    "layer_norm_style": "pre",
+                    "multi_head_config": {
+                        "num_heads": self.hparams.n_head,
+                        "residual_dropout": self.hparams.resid_pdrop,
+                        "use_rotary_embeddings": True,
+                        "attention": {
+                            "name": self.hparams.attention,
+                            "dropout": self.hparams.attn_pdrop,
+                            "causal": True,
```

Comment:
For causal to be enabled we need to pass a seq_len right? I wasn't sure what to set the seq len (from what I gather, we need the infered size after the conv layer)

Reply:
ah it depends on the attentions, some of them can do without a seq len (which is really a 'max seq len'). In that case it should just be the num_patches + class token I think, I can check that out